### ISSUE TITLE
Update API for PSK to enable TLS 1.3

### ISSUE THREAD

**Issue by daverodgman** [ARMmbed/mbedtls#4121]
For TLS 1.3, it would be helpful to make the PSK config part of the SSL context, rather than global configuration.

**Comment by mpg** [comment]
In general, there's 3 ways a particular SSL setting can be configured (in order of precedence):

1. per handshake, using a `mbedtls_ssl_set_hs_xxx()` function inside a callback during the handshake
2. per context, using a `mbedtls_ssl_set_xxx()` function before starting the handshake
3. shared config, using a `mbedtls_ssl_conf_xxx()` function before any handshake

For most settings we have either (2) or (3), and for those where a callback makes sense we have also (1) in addition. For PSK, currently we have (1) server-side (client-side has a community PR pending: #3831) and (3). 

If I understand correctly, the request is to remove (3) and add (2) instead. My question is: do we really need to remove (3)? Can't we keep it and just add (2)? I have two things in mind with this question:

- Is this something that really needs to be tied to 3.0, or could it be done at another time (without too much additional cost)?
- We try to choose between (2) and (3) based on how we expect the setting will be used most of the time: is it likely to be shared between multiple connection? If yes, it goes in the config (3), otherwise in the context (2). Apparently we thought that for TLS <= 1.2, shared config was the right choice. I understand that for TLS 1.3, per-context will be needed, so that's a reason to add a per-context setting, but is that a sufficient reason to remove the shared config? (I appreciate having multiple ways to set somethings comes at a cost in terms of code size, testing and documentation, so I can see that reason, but I'm asking if that means having the shared config really blocks adding a per-context setting.)

Cc-ing @hanno-arm as this is related to TLS 1.3.

**Comment by hanno-arm** [comment]
@mpg Thank you for your feedback, and apologies for not replying earlier - I missed the CC.

We have to consider that in TLS 1.3, resumptions based on PSKs, Session IDs, and Tickets, are all technically the same and it would therefore be ideal if they'd be configured through the same means in the Mbed TLS API. In Mbed TLS 2.X, however, Session ID and Ticket Based resumption are configured on a per-context level through `mbedtls_ssl_session_{get,set}()`, while PSK based resumption is configured on a per-config level.

In the outline for how to consolidate those resumption mechanisms that I wrote on the [mailing list](https://lists.trustedfirmware.org/pipermail/mbed-tls/2020-April/000049.html) some time ago, the idea was to make PSK-configuration a convenience wrapper around `mbedtls_ssl_session_set()` which would construct a ticket/session (there's no distinction anymore in the proposal between those concepts) structure from the provided PSK and load that via `mbdtls_ssl_session_set()`. This seems easier if PSK configuration applies to the SSL context, not the SSL configuration. However, one could defer the translation of `mbedtls_ssl_conf_psk()` -- applying to an SSL configuration -- into `mbedtls_ssl_session_set()` -- applying to an SSL context -- to the point where an SSL context has been bound to the configuration and is about to do a handshake.

It also needs to be considered that PSK-based resumption is more dynamic in TLS 1.3 than it is in TLS 1.2: The use of PSKs in TLS 1.2 is what's called "external PSK" in TLS 1.3. Those PSKs are pre-established out of band and it's thus not unreasonable to consider them static and track them in the SSL configuration.  However, TLS 1.3 _resumption_ PSKs (= tickets) are dynamic and arise as you establish connections. You cannot make use of those PSKs in the same way as TLS 1.2 tickets if they are configured through the config, as you're not allowed to modify the SSL configuration while SSL contexts are bound to it.

What we could do is keep `mbedtls_ssl_conf_psk()` as a convenience function for the static configuration of _external_ PSKs only, and, as mentioned above, translate it into a context-specific call to `mbedtls_ssl_session_set()` once an SSL context is about to do a handshake.

This approach has some limitations:
* We inherit the limitation that you can only configure one PSK. In fact, thinking about it, I think this is something we should probably change for Mbed TLS 3 regardless of the TLS 1.3 considerations: Instead of prescribing that the implementation stores exactly one PSK and further calls to `mbedtls_ssl_conf_psk()` will overwrite earlier ones, we might want to leave this implementation-defined and only specify which error code will be returned once the implementation cannot store any further PSKs. I will open an independent issue for this.
* In TLS 1.3, there's essential metadata attached to any PSK which cannot be configured through `mbedtls_ssl_conf_psk()` as-is because it only PSK Identity + PSK. This metadata includes: (a) the lifetime of the PSK, (b) whether it may be used for 0-RTT, and if so, for how much, (c) the KDF to be used when resuming a session based on this PSK. If we don't let the user choose, we have to prescribe default values, which isn't optimal. 
* Continuing the previous point, external PSKs in TLS 1.2 are not bound to a specific KDF. The RFC indicates that implementations may want to treat PSKs for TLS 1.2 and TLS 1.3 independently.

After all this, I'm currently inclined to consider having separate configuration functions for external PSKs for TLS <= 1.2 and TLS 1.3, restricting `mbedtls_ssl_conf_psk()` to TLS <= 1.2 when we add support for TLS 1.3.

**Comment by mpg** [comment]
@hanno-arm Thanks for you response! I think it's now clearer to me why having the PSK set in the context rather than the config would makes things more uniform with how tickets/session-IDs will be managed. However, I seem to understand that if this were to miss the boat for 3.0, this wouldn't entirely block us and we'd still have options available. Its that correct?

**Comment by hanno-arm** [comment]
@mpg Yes, I think that's right - we could always qualify `mbedtls_ssl_conf_psk()` as _external PSKs only_, in which case it seems right that it's bound to the config, not the context. 

What's more important for 3.0 is to think through how to handle resumption PSKs aka tickets aka sessions as tracked by https://github.com/ARMmbed/mbedtls/issues/4184.

**Comment by daverodgman** [comment]
@ronald-cron-arm is this still relevant?

**Comment by daverodgman** [comment]
Discussed with RC - agreed still required. This could be done after 4.0 with no API-break by adding a function like `mbedtls_ssl_conf_psk_ext()` (but obviously it would be nicer to avoid this).
